zkvm/docs/zkvm-blockchain.md

```diff
@@ -0,0 +1,406 @@
+[ZkVM transactions](zkvm-spec.md).
+
+Nodes participating in a ZkVM blockchain network must implement the data types and perform the procedures described in this document.
+Specifically:
+
+- Each node maintains a [blockchain state](#blockchain-state).
+- A node creating a new ZkVM network performs the [start new network](#start-new-network) procedure.
+- A node joining an existing ZkVM network performance the [join existing network](#join-existing-network) procedure.
+- Each node performs the [apply block](#apply-block) procedure on the arrival of each new block.
+
+This document does not describe the mechanism for producing blocks from pending transactions,
+nor for choosing among competing blocks to include in the authoritative chain
+(a.k.a. consensus).
+
+_TBD: add a consensus spec._
+
+# Data types
+
+## Blockchain state
+
+The state of a ZkVM blockchain is given by the blockchain-state structure.
+Each node maintains a copy of this structure.
+As each new
+[block](#block)
+arrives,
+it is
+[applied](#apply-block)
+to the current state to produce a new,
+updated state.
+
+The blockchain state contains:
+
+- `initialheader`:
+  The initial block header
+  (the header with height 1).
+  This never changes.
+- `tipheader`:
+  The latest block header.
+- `utxos`:
+  The set of current
+  [utxo IDs](zkvm-spec.md#utxo).
+- `nonces`:
+  The set of current
+  [nonce anchors](zkvm-spec.md#nonce).
+- `refids`:
+  The set of recent
+  [block IDs](#block-id).
+  The size of this set is bounded by `block.header.refscount`.
+  Block IDs in this set may be referenced by nonces.
+
+## Block
+
+A block contains:
+
+- `header`:
+  a
+  [block header](#block-header).
+- `txs`:
+  a list of
+  [transactions](zkvm-spec.md#transaction).
+
+The initial block
+(at height 1)
+has an empty list of transactions.
+
+## Block header
+
+A block header contains:
+
+- `version`:
+  integer version number,
+  set to 1.
+- `height`:
+  integer block height.
+  Initial block has height 1.
+  Height increases by 1 with each new block.
+- `previd`:
+  ID of the preceding block.
+  For the initial block
+  (which has no predecessor),
+  this is an all-zero string of 32 bytes.
+- `timestamp_ms`:
+  integer timestamp of the block in milliseconds since the Unix epoch:
+  00:00:00 UTC Jan 1,
+  1970.
+  Each new block must have a time strictly later than the block before it.
+- `txroot`:
+  32-byte
+  [Merkle root hash](zkvm-spec.md#merkle-binary-tree)
+  of the transactions in the block.
+- `utxoroot`:
+  32-byte
+  [Merkle patricia root hash](#merkle-patricia-tree)
+  of the utxo set after applying all transactions in the block.
+- `nonceroot`:
+  32-byte
+  [Merkle patricia root hash](#merkle-patricia-tree)
+  of the nonce set after applying all transactions in the block.
+- `refscount`:
+  integer number of recent block IDs to store for reference.
+  A new block may specify a lower `refscount` than its predecessor but may not increase it by more than 1.
```

Comment:
let's add `ext` variable-length field that we can use to throw in extra fields later as a soft-fork? And also hash it into block-id below.

Reply:
TxVM defined the blockheader as a TxVM tuple with arbitrary number of extra fields, that were all hashed into block ID. Here we can use varlength string and extend it how we want later. The only caveat is that we'd need to hash it in as a whole for compatibility, even if we break it into separate fields in future updates.